actors/builtin/market/market_balances.go

```diff
@@ -0,0 +1,115 @@
+package market
+
+import (
+	"errors"
+	"fmt"
+
+	addr "github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+	"github.com/filecoin-project/specs-actors/actors/runtime/exitcode"
+	. "github.com/filecoin-project/specs-actors/actors/util"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"golang.org/x/xerrors"
+)
+
+func (st *State) lockBalanceOrAbort(et, lt *adt.BalanceTable, proposal *DealProposal) (error, exitcode.ExitCode) {
+	err, code := st.maybeLockBalance(et, lt, proposal.Client, proposal.ClientBalanceRequirement())
+	if err != nil {
+		return fmt.Errorf("failed to lock client funds: %w", err), code
+	}
+
+	err, code = st.maybeLockBalance(et, lt, proposal.Provider, proposal.ProviderCollateral)
+	if err != nil {
+		return fmt.Errorf("failed to lock provider funds: %w", err), code
+	}
+
+	st.TotalClientLockedCollateral = big.Add(st.TotalClientLockedCollateral, proposal.ClientCollateral)
```

Comment:
Far out idea that you shouldn't do in this PR but could think about: instead of making these state methods you could pass in pointers to relevant token amounts etc that live on the state mutation object the same way the adts do.  During commit operation you could then write these things to the state object

Reply:
If methods on the State object do the commit in the same method call tree as the writes, I think this is ok. The state object can be entirely responsible for its internal consistency. But if correctness depends on the user of the state object remembering to flush it, it's less good.

Reply:
@anorth @ZenGround0 Have address this in the updated PR. These locked balances will now be flushed only when the corresponding HAMTs are flushed.